### ISSUE TITLE
X-Forwarded-Port overrides X-Forwarded-For

### ISSUE THREAD

**Issue by stuban** [eclipse/jetty.project#3782]
For us the changes related to #3630 have some surprising regression with 9.4.19
Our reverse-proxy (haproxy) always sends both X-Forwarded-Port and X-Forwarded-Host headers towards jetty.
This always worked fine but with 9.4.19 it looks like the X-Forwarded-Port is taken as host.
A simple example just printing out `request.getRequestURL()`
Request forwarded from upstream haproxy:
```
GET /test/forwardedTest HTTP/1.1
Host: linux-0py3.test-example.de
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en,en-US;q=0.95,de;q=0.89,fr-FR;q=0.84,nl;q=0.79,de-DE;q=0.74,en-GB;q=0.68,de-US;q=0.63,en-EN;q=0.58,fr;q=0.53,nl-NL;q=0.47,de-formal;q=0.42,en-CA;q=0.37,en-AU;q=0.32,es-ES;q=0.26,es;q=0.21,en-DE;q=0.16,eng-US;q=0.11,eng;q=0.05
Accept-Encoding: gzip, deflate
DNT: 1
Upgrade-Insecure-Requests: 1
Cache-Control: max-age=0
X-Forwarded-Port: 80
X-Forwarded-For: 192.168.251.50
```

Answer from 9.4.18:
```
HTTP/1.1 200 OK
Content-Type: text/html;charset=utf-8
Content-Length: 179

<html lang="en"><head><title>Servlet Hello</title></head>
<body><h2>Hello World!</h2>
<p>As seen by jetty: http://linux-0py3.test-example/test/forwardedTest</p>
</body></html>
```

Answer from 9.4.19:
```
HTTP/1.1 200 OK
Content-Type: text/html;charset=utf-8
Content-Length: 155

<html lang="en"><head><title>Servlet Hello</title></head>
<body><h2>Hello World!</h2>
<p>As seen by jetty: http://80/test/forwardedTest</p>
</body></html>
```

**Comment by joakime** [comment]
`X-Forwarded-Port` is a horrible header.
It's not a standard, and everyone uses it for different purposes.
Mozilla, the original author of the `X-Forwarded-*` headers does not recognize it.

1. Some use it to represent the Remote Address port (aka, the Client Socket local port)
2. Some use it to represent the Server Port used by the client (aka the Load Balancer Socket listening port)
3. Most servers refuse to use it entirely.

We get reports from users expecting (1) and we fix it for (2).
Then we get reports from users expecting (2) and we fix it for (1).
Right now, with Jetty 9.4.19, the code is using as (1).

The formalized standard for Forwarding headers, https://tools.ietf.org/html/rfc7239 mentions the transition of the `X-Forwarded-*` headers, but not `X-Forwarded-Port`, which doesn't exist as a standard anywhere.
Of the two port concepts seen in use on the open internet, they are clearly defined in the formalized RFC7239 spec as `by=<client-host:client-port>`, and `for=<node-host:node-port>` values on the `Forwarded` header ("node" is a load balancer / firewall / etc ...)

Of note, [Amazon AWS seems to use this `X-Forwarded-Port` header](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/x-forwarded-headers.html), and they use it as type (2).
But even Amazon AWS is aggressively transitioned away to using RFC7239 exclusively. 

I don't think our usage as type (1) is correct, seeing as we've been using it as a type (2) for an even longer time.
I'll start a branch and see about going back to type (2) usage for it.

**Comment by joakime** [comment]
Opened PR #3784

**Comment by stuban** [comment]
Thanks, I know the x-forwarded topic is quite annoying. As mentioned it was more a kind of surprising regression. I can change the settings on the haproxy where I'm in control of it. Does jetty already support RFC7239 style?
But unfortunately there are as well some reverse proxies beyond our control and the  X-Forwarded-Port is the only way to distinguish between HTTP and HTTPS there.

**Comment by gregw** [comment]
@stuban How do you have the customizer configured?  I've tried to reproduce your result, but with the request:
```java
        String response = _connector.getResponse(
            "GET /test HTTP/1.1\n" +
                "Host: myhost.com\n" +
                "X-Forwarded-Port: 8888\n" +
                "X-Forwarded-For: 10.9.8.7\n" +
                "\n");
```
I'm seeing `getRequestURL` as `http://myhost.com/test`, not as `http://8888/test` as your report suggests!

I still think we are wrong and it should be `http://myhost.com:8888/test`, but not wrong in the same way that you report?

**Comment by gregw** [comment]
@stuban also note that a work around for you maybe to call `setForwardedPortHeader(null)` to disable handling of that header.

**Comment by joakime** [comment]
@stuban yes, Jetty supports RFC7239 style already.
You don't have to change anything about the `ForwardedRequestCustomizer` to start using it.

But note that there is an optional setting `ForwardedRequestCustomizer.setForwardedOnly(true);` which will make it only support the RFC7239 `Forwarded` header, ignoring all others legacy forwarding headers.

**Comment by stuban** [comment]
@gregw It seems to depend on the setup. When running a simple self contained / embedded testcase i.e. via mvn jetty:run I can not reproduce either.
But It happens when deployed to a standalone jetty instance with "http-forwarded" module enabled.
````
Enabled Modules:
================
    0) mail            transitive provider of mail for jndi
    1) bytebufferpool  transitive provider of bytebufferpool for server
                       init template available with --add-to-start=bytebufferpool
    2) threadpool      transitive provider of threadpool for server
                       init template available with --add-to-start=threadpool
    3) server          transitive provider of server for http
                       transitive provider of server for plus
                       transitive provider of server for security
                       transitive provider of server for servlet
                       transitive provider of server for requestlog
                       transitive provider of server for jndi
                       init template available with --add-to-start=server
    4) jndi            transitive provider of jndi for plus
    5) security        transitive provider of security for webapp
                       transitive provider of security for plus
    6) transactions    transitive provider of transactions for plus
    7) servlet         transitive provider of servlet for webapp
                       transitive provider of servlet for jsp
    8) webapp          transitive provider of webapp for plus
                       transitive provider of webapp for deploy
                       init template available with --add-to-start=webapp
    9) plus            ${jetty.base}/start.d/plus.ini
   10) annotations     transitive provider of annotations for websocket
                       transitive provider of annotations for jsp
   11) apache-jsp      transitive provider of apache-jsp for jsp
   12) client          transitive provider of client for websocket
   13) console-capture ${jetty.base}/start.d/console-capture.ini
   14) deploy          ${jetty.base}/start.d/deploy.ini
   15) http            ${jetty.base}/start.d/http.ini
   16) http2c          ${jetty.base}/start.d/http2c.ini
   17) http-forwarded  ${jetty.base}/start.d/http-forwarded.ini
   18) jsp             ${jetty.base}/start.d/jsp.ini
   19) requestlog      ${jetty.base}/start.d/requestlog.ini
   20) websocket       ${jetty.base}/start.d/websocket.ini
````
No customization in http-forwarded.ini

Thanks for the suggestions we will try to move towards the RFC7239

**Comment by joakime** [comment]
PR #3784 has been merged into `jetty-9.4.x` branch

**Comment by joakime** [comment]
Closing, as this is already merged and available in Jetty 9.4.20
